experiment/adscert/signer.go

```diff
@@ -0,0 +1,129 @@
+	req := &api.AuthenticatedConnectionSignatureRequest{
+		RequestInfo: createRequestInfo(destinationURL, body),
+	}
+	resp, err := ips.signatory.SignAuthenticatedConnection(req)
+	if err != nil {
+		return "", err
+	}
+	if resp.GetSignatureOperationStatus() == api.SignatureOperationStatus_SIGNATURE_OPERATION_STATUS_OK {
+		signatureMessage := resp.RequestInfo.SignatureInfo[0].SignatureMessage
+		return signatureMessage, nil
+	}
+	return "", fmt.Errorf("error signing request: %s", resp.GetSignatureOperationStatus())
+}
+
+func newInProcessSigner(inProcessSignerConfig config.InProcess) *inProcessSigner {
+	return &inProcessSigner{
+		signatory: signatory.NewLocalAuthenticatedConnectionsSignatory(
+			inProcessSignerConfig.Origin,
+			rand.Reader,
+			clock.New(),
+			discovery.NewDefaultDnsResolver(),
+			discovery.NewDefaultDomainStore(),
+			time.Duration(inProcessSignerConfig.DNSCheckIntervalInSeconds)*time.Second,
+			time.Duration(inProcessSignerConfig.DNSRenewalIntervalInSeconds)*time.Second,
+			[]string{inProcessSignerConfig.PrivateKey}),
+	}
+}
+
+func createRequestInfo(destinationURL string, body []byte) *api.RequestInfo {
+	reqInfo := &api.RequestInfo{}
+	signatory.SetRequestInfo(reqInfo, destinationURL, body)
+	return reqInfo
+}
+
+type remoteSigner struct {
+	signatory signatory.AuthenticatedConnectionsSignatory
+}
+
+func (rs *remoteSigner) Sign(destinationURL string, body []byte) (string, error) {
+	// The RequestInfo proto contains details about the individual ad request
+	// being signed.  A SetRequestInfo helper function derives a hash of the
+	// destination URL and body, setting these value on the RequestInfo message.
+	reqInfo := &api.RequestInfo{}
+	signatory.SetRequestInfo(reqInfo, destinationURL, []byte(body))
+
+	// Request the signature.
+	signatureResponse, err := rs.signatory.SignAuthenticatedConnection(
+		&api.AuthenticatedConnectionSignatureRequest{
+			RequestInfo: reqInfo,
+		})
+	if err != nil {
+		return "", err
+	}
+	if signatureResponse != nil && signatureResponse.SignatureOperationStatus == api.SignatureOperationStatus_SIGNATURE_OPERATION_STATUS_OK {
+		signatureMessage := signatureResponse.RequestInfo.SignatureInfo[0].SignatureMessage
+		return signatureMessage, err
+	}
+	return "", fmt.Errorf("error signing request: %s", signatureResponse.GetSignatureOperationStatus())
+}
+
+func newRemoteSigner(remoteSignerConfig config.Remote) (*remoteSigner, error) {
+	// Establish the gRPC connection that the client will use to connect to the
+	// signatory server.  This basic example uses unauthenticated connections
+	// which should not be used in a production environment.
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+	conn, err := grpc.Dial(remoteSignerConfig.Url, opts...)
+	if err != nil {
+		fmt.Errorf("failed to dial remote signer: %v", err)
+	}
+	//defer conn.Close() -- where this should be?
```

Comment:
Ideally, this should be triggered when the app is gracefully shutting down. You might be able to just get away without it though, since the connections are expected to remain open for the life of the application.

Reply:
Right, this should live and die with application. I removed this line :/